// playcanvas/build/playcanvas.dbg/src/framework/app-options.js
/**
 * @import { BatchManager } from '../scene/batching/batch-manager.js'
 * @import { ComponentSystem } from './components/system.js'
 * @import { ElementInput } from './input/element-input.js'
 * @import { GamePads } from '../platform/input/game-pads.js'
 * @import { GraphicsDevice } from '../platform/graphics/graphics-device.js'
 * @import { Keyboard } from '../platform/input/keyboard.js'
 * @import { Lightmapper } from './lightmapper/lightmapper.js'
 * @import { Mouse } from '../platform/input/mouse.js'
 * @import { ResourceHandler } from './handlers/handler.js'
 * @import { SoundManager } from '../platform/sound/manager.js'
 * @import { TouchDevice } from '../platform/input/touch-device.js'
 * @import { XrManager } from './xr/xr-manager.js'
 */

/**
 * AppOptions is an object that holds configuration settings utilized in the creation of AppBase. It
 * allows functionality to be included or excluded from the AppBase instance.
 */
class AppOptions {
  constructor() {
    /**
     * Input handler for {@link ElementComponent}s.
     *
     * @type {ElementInput}
     */
    this.elementInput = void 0;
    /**
     * Keyboard handler for input.
     *
     * @type {Keyboard}
     */
    this.keyboard = void 0;
    /**
     * Mouse handler for input.
     *
     * @type {Mouse}
     */
    this.mouse = void 0;
    /**
     * TouchDevice handler for input.
     *
     * @type {TouchDevice}
     */
    this.touch = void 0;
    /**
     * Gamepad handler for input.
     *
     * @type {GamePads}
     */
    this.gamepads = void 0;
    /**
     * Prefix to apply to script urls before loading.
     *
     * @type {string}
     */
    this.scriptPrefix = void 0;
    /**
     * Prefix to apply to asset urls before loading.
     *
     * @type {string}
     */
    this.assetPrefix = void 0;
    /**
     * Scripts in order of loading first.
     *
     * @type {string[]}
     */
    this.scriptsOrder = void 0;
    /**
     * The sound manager
     *
     * @type {SoundManager}
     */
    this.soundManager = void 0;
    /**
     * The graphics device.
     *
     * @type {GraphicsDevice}
     */
    this.graphicsDevice = void 0;
    /**
     * The lightmapper.
     *
     * @type {typeof Lightmapper}
     */
    this.lightmapper = void 0;
    /**
     * The BatchManager.
     *
     * @type {typeof BatchManager}
     */
    this.batchManager = void 0;
    /**
     * The XrManager.
     *
     * @type {typeof XrManager}
     */
    this.xr = void 0;
    /**
     * The component systems the app requires.
     *
     * @type {typeof ComponentSystem[]}
     */
    this.componentSystems = [];
    /**
     * The resource handlers the app requires.
     *
     * @type {typeof ResourceHandler[]}
     */
    this.resourceHandlers = [];
  }
}

export { AppOptions };
